File: packages/shared/src/lib/format.ts
export function formatCOP(value: number | null | undefined): string {
  if (value == null) return '\u2014';
  return new Intl.NumberFormat('es-CO', {
    style: 'currency',
    currency: 'COP',
    minimumFractionDigits: 0,
    maximumFractionDigits: 0,
  }).format(value);
}

export function formatCOPCompact(value: number | null | undefined): string {
  if (value == null) return '\u2014';
  if (value >= 1_000_000) return `$${(value / 1_000_000).toFixed(1)}M`;
  if (value >= 1_000) return `$${(value / 1_000).toFixed(0)}K`;
  return `$${value.toFixed(0)}`;
}

export function formatPctChange(value: number | null | undefined): string {
  if (value == null) return '\u2014';
  const sign = value >= 0 ? '+' : '';
  return `${sign}${value.toFixed(1)}%`;
}

export function formatKg(value: number | null | undefined): string {
  if (value == null) return '\u2014';
  if (value >= 1_000_000) return `${(value / 1_000_000).toFixed(1)}M kg`;
  if (value >= 1_000) return `${(value / 1_000).toFixed(1)}K kg`;
  return `${value.toFixed(0)} kg`;
}

export function formatDateShort(dateStr: string): string {
  const months = ['ene', 'feb', 'mar', 'abr', 'may', 'jun', 'jul', 'ago', 'sep', 'oct', 'nov', 'dic'];
  const d = new Date(dateStr + 'T00:00:00');
  return `${d.getDate()} ${months[d.getMonth()]}`;
}

export function formatDateMedium(dateStr: string): string {
  const months = ['enero', 'febrero', 'marzo', 'abril', 'mayo', 'junio', 'julio', 'agosto', 'septiembre', 'octubre', 'noviembre', 'diciembre'];
  const d = new Date(dateStr + 'T00:00:00');
  return `${d.getDate()} de ${months[d.getMonth()]} ${d.getFullYear()}`;
}

export function formatPriceContext(presentation?: string | null, units?: string | null): string {
  const parts = [presentation, units].filter(Boolean);
  return parts.join(' \u00b7 ');
}

export function pctChange(oldVal: number, newVal: number): number {
  if (oldVal === 0) return 0;
  return ((newVal - oldVal) / oldVal) * 100;
}

File: agroamigo-iphone/lib/services/format.dart
import 'package:intl/intl.dart';

/// Port of `agroamigo-app/src/lib/format.ts`.

final _copFormatter = NumberFormat.currency(
  locale: 'es_CO',
  symbol: r'$',
  decimalDigits: 0,
);

String formatCOP(num? value) {
  if (value == null) return '—';
  return _copFormatter.format(value);
}

String formatCOPCompact(num? value) {
  if (value == null) return '—';
  if (value >= 1000000) return '\$${(value / 1000000).toStringAsFixed(1)}M';
  if (value >= 1000) return '\$${(value / 1000).round()}K';
  return '\$${value.round()}';
}

String formatPctChange(num? value) {
  if (value == null) return '—';
  final sign = value >= 0 ? '+' : '';
  return '$sign${value.toStringAsFixed(1)}%';
}

String formatKg(num? value) {
  if (value == null) return '—';
  if (value >= 1000000) return '${(value / 1000000).toStringAsFixed(1)}M kg';
  if (value >= 1000) return '${(value / 1000).toStringAsFixed(1)}K kg';
  return '${value.round()} kg';
}

const _shortMonthsEs = ['ene', 'feb', 'mar', 'abr', 'may', 'jun',
  'jul', 'ago', 'sep', 'oct', 'nov', 'dic'];

const _longMonthsEs = ['enero', 'febrero', 'marzo', 'abril', 'mayo', 'junio',
  'julio', 'agosto', 'septiembre', 'octubre', 'noviembre', 'diciembre'];

String formatDateShort(String dateStr) {
  final d = DateTime.parse('${dateStr}T00:00:00');
  return '${d.day} ${_shortMonthsEs[d.month - 1]}';
}

String formatDateMedium(String dateStr) {
  final d = DateTime.parse('${dateStr}T00:00:00');
  return '${d.day} de ${_longMonthsEs[d.month - 1]} ${d.year}';
}

String formatPriceContext(String? presentation, String? units) {
  return [presentation, units].where((p) => p != null && p.isNotEmpty).join(' · ');
}

double pctChange(num oldVal, num newVal) {
  if (oldVal == 0) return 0;
  return ((newVal - oldVal) / oldVal) * 100;
}
